consensus/hotstuff/votecollector/statemachine.go

```diff
@@ -55,129 +73,174 @@ func NewStateMachine(base CollectionBase) *StateMachine {
 // delegate function call, otherwise we will return an error indicating wrong collector state.
 // ATTENTION: this might be changed if CreateVote and state transitions will be called in parallel
 // something like compare-and-repeat might need to be implemented.
-func (m *StateMachine) CreateVote(block *model.Block) (*model.Vote, error) {
-	collector := m.atomicLoadCollector()
-	blockSigner, ok := collector.(hotstuff.BlockSigner)
+func (m *VoteCollector) CreateVote(block *model.Block) (*model.Vote, error) {
+	processor := m.atomicLoadProcessor()
+	blockSigner, ok := processor.(hotstuff.BlockSigner)
 	if ok {
 		return blockSigner.CreateVote(block)
 	}
 	return nil, ErrDifferentCollectorState
 }
 
-func (m *StateMachine) AddVote(vote *model.Vote) error {
+func (m *VoteCollector) AddVote(vote *model.Vote) error {
+	// Cache vote
+	err := m.votesCache.AddVote(vote)
```

Comment:
Do we still cache the vote even if we have received the block?

Reply:
We want to keep _all_ received votes, so we can do forensics later (and scan for double voting across all votes for this view).